client/src/components/RecipeShow.js

```diff
@@ -11,26 +16,140 @@ const RecipeShow = (props) => {
+      const body = await response.json()
+      setPantryIngredients(body.ingredients)
+    } catch(err) {
+      console.error(`Error in fetch: ${err.message}`)
+    }
+  }
+
   useEffect(() => {
       fetchRecipeDetails()
-  }, [])
+      fetchPantryIngredients()
+    }, [])
 
-  let ingredientsList
-  if(recipe.ingredients) {
-    ingredientsList = recipe.ingredients.map(ingredient => {
+    useEffect(() => {
+      if(recipeIngredients.length > 0) {
+        getNewTotals()
+      }
+    }, [recipeIngredients, pantryIngredients])
+
+    const getNewTotals = () => {
+      let ingredientsToUpdate = recipeIngredients.map(ingredient => {
+        const ingredientInPantry = pantryIngredients.find(pantryIngredient => pantryIngredient.name.toLowerCase() === ingredient.name)
+        if(ingredientInPantry) {
+          if(ingredientInPantry.unit === ingredient.unit) {
+            const updatedIngredient = {
+              ...ingredientInPantry,
+              amount: ingredientInPantry.amount - ingredient.amount
+            }
+            return updatedIngredient
+            } else {
+            const convertedRecipeIngredient = Ingredients.convertIngredient(ingredient)
+            const convertedPantryIngredient = Ingredients.convertIngredient(ingredientInPantry)
+            if (!convertedRecipeIngredient || !convertedPantryIngredient) {
+              return false
+            } else {
+              const newTotal = Ingredients.convertAndUpdate(ingredientInPantry, ingredient)
+              const updatedIngredient = {
+                ...ingredientInPantry,
+                amount: newTotal
+              }
+              return updatedIngredient
+            }
+          }
+        } else {
+          return false
+        }
+      })
+      setUpdatedIngredients(ingredientsToUpdate)
+    }
+    
+    const compareIngredient = (ingredient) => {
+      const ingredientInPantry = pantryIngredients.find(pantryIngredient => pantryIngredient.name.toLowerCase() === ingredient.name)
+      if(ingredientInPantry) {
+        if(ingredientInPantry.unit === ingredient.unit) {
+          if(ingredientInPantry.amount >= ingredient.amount) {
+            return "have enough"
+          } else {
+            return "need more"
+          }
+        } else {
+          const convertedRecipeIngredient = Ingredients.convertIngredient(ingredient)
+          const convertedPantryIngredient = Ingredients.convertIngredient(ingredientInPantry)
+          if (!convertedRecipeIngredient || !convertedPantryIngredient) {
+            return "can't tell"
+          } else {
+            if (convertedPantryIngredient - 3 >= convertedRecipeIngredient) {
+              return "have enough"
+            } else if (convertedPantryIngredient + 3 <= convertedRecipeIngredient) {
+              return "need more"
+            } else {
+              return "it's close"
+            }
+          }
+        }
+      } else {
+        return "don't have"
+      }
+    }
```

Comment:
same thing for this. Awesome job, but we can move this into a POJO. And some if not all of it can be computed on the backend

Reply:
particularly the logic in the second else statement